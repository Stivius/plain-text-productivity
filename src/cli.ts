#! /usr/bin/env node

import main from "./index";

main().catch((err) => console.log(err, "Error while running"));
