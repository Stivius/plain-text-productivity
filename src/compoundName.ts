const SEPARATOR = "/";

// if has children than can't be used as record
// EstimatedProject contain only leafs

export class CompoundName {
  parent: CompoundName;
  name: string;

  constructor(value: string) {
    const splitted = value.split(SEPARATOR);
    if (splitted.length > 0) {
      const name = splitted.pop();
      if (name.length === 0) {
        throw Error("subpart cannot be empty");
      }
      if (splitted.length > 0) {
        this.parent = new CompoundName(splitted.join(SEPARATOR));
      }
      this.name = name;
    } else {
      this.name = value;
    }
  }

  toString(): string {
    if (this.parent != null) {
      return `${this.parent.toString()}/${this.name}`;
    }
    return this.name;
  }

  isEqual(other: CompoundName): boolean {
    return this.toString() === other.toString();
  }
}

export const searchForCompoundProject = (searchedValue: CompoundName) => {
  return (value: CompoundName) => value.isEqual(searchedValue);
};
