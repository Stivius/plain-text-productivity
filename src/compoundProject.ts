const SEPARATOR = '/'

// if has children than can't be used as record
// EstimatedProject contain only leafs

export class CompoundProject {
    parent: CompoundProject;
    name: string;

    constructor(value: string) {
        const splitted = value.split(SEPARATOR);
        if (splitted.length > 0) {
            const name = splitted.pop();
            if (name.length === 0) {
                throw Error('subpart cannot be empty');
            }
            if (splitted.length > 0) {
                this.parent = new CompoundProject(splitted.join(SEPARATOR));
            }
            this.name = name;
        } else {
            this.name = value;
        }
    }

    toString(): string {
        if (this.parent != null) {
            return `${this.parent.toString()}/${this.name}`
        }
        return this.name;
    }

    isEqual(other: CompoundProject): boolean {
        return this.toString() === other.toString();
    }
}
