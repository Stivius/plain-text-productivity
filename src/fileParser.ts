import fs from 'fs';
import moment from 'moment';
import {
    FileData,
    Record,
    Mark,
    EstimatedProject,
    DATE_REGEX,
    PROJECT_NAME_REGEX,
    METADATA_PROJECT_NAME_REGEX
} from "./interfaces";

enum ParsingState {
    Init,
    MetadataStarted,
    MetadataActiveProjects,
    ArchivedProjects,
    Blank,
    Record,
}

export class FileParser {
    private projects: string[] = [];
    private records: Record[] = [];
    private currentRecord: Record = undefined;

    constructor(private filename: string) { }

    parse(): FileData {
        const data = fs.readFileSync(this.filename);
        const lines = data.toString().split('\n');

        const METADATA_SEPARATOR = '---';

        let state: ParsingState = ParsingState.Init;
        let metadataLines = 0;

        for (const line of lines) {
            const trimmedLine = line.trim();
            ++metadataLines;
            // TODO: state machine
            if (trimmedLine === METADATA_SEPARATOR) {
                if (state === ParsingState.Init) {
                    state = ParsingState.MetadataStarted;
                } else {
                    state = ParsingState.Blank;
                }
            }
            else if (trimmedLine.indexOf('Projects:') !== -1) {
                if (state === ParsingState.MetadataStarted) {
                    state = ParsingState.MetadataActiveProjects;
                }
            }
            else if (state === ParsingState.MetadataActiveProjects) {
                this.projects.push(this.parseMetadataProject(trimmedLine));
            }
            else if (trimmedLine.length === 0 && state !== ParsingState.Init) {
                state = ParsingState.Blank;
                if (this.currentRecord !== undefined) {
                    this.records.push(this.currentRecord);
                    this.currentRecord = undefined;
                }
            }
            else if (trimmedLine.match(DATE_REGEX) && state === ParsingState.Blank) {
                state = ParsingState.Record;
                this.currentRecord = this.createRecord(trimmedLine);
            }
            else if (state === ParsingState.Record) {
                this.currentRecord.projects.push(this.parseProject(trimmedLine));
            }
        }
        return { metadata: { projects: this.projects }, records: this.records };
    }

    private createRecord(line: string) {
        if (!moment(line, 'YYYY-MM-DD').isValid()) {
            throw Error('invalid date')
        }
        return {
            day: new Date(line),
            projects: [],
        }
    }

    private parseMetadataProject(line: string) {
        if (!line.match(METADATA_PROJECT_NAME_REGEX)) {
            throw Error('invalid metadata project');
        }
        const parsedProject = line.replace('-', '').trim();
        if (this.projects.indexOf(parsedProject) !== -1) {
            throw Error('duplicated metadata project');
        }
        return parsedProject;
    }

    private parseProject(line: string): EstimatedProject {
        const splitted = line.split(':');
        const [ name, mark ] = splitted;
        if (splitted.length !== 2 || this.currentRecord === undefined || !name.match(PROJECT_NAME_REGEX)) {
            throw new Error('invalid record')
        }
        if (this.currentRecord.projects.find((value) => value.name === name) !== undefined) {
            throw new Error('duplicated project record')
        }
        return {
            name: splitted[0],
            mark: this.parseMark(mark),
        };
    }

    private parseMark(value: string): Mark {
        // TODO: mark regex validate
        const trimmedValue = value.trim();
        if (trimmedValue === '-') {
            return undefined;
        }
        const mark = parseInt(value);
        if (Number.isNaN(mark) || mark < 1 || mark > 5) {
            throw new Error('invalid mark')
        }
        return mark as Mark;
    }
}
