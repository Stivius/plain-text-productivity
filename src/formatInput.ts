import chalk from "chalk";
import columnify from "columnify";
import dateFormat from "dateformat";
import prompt from "prompt";

function formatDate(date: Date) {
  return dateFormat(date, "yyyy-mm-dd");
}

export function formatInputForFile(date: Date, data: prompt.Properties) {
  return `${formatDate(date)}\n`
    .concat(
      Object.keys(data)
        .map((k) => {
          return `${k}:${data[k]}`;
        })
        .join("\n")
    )
    .concat("\n\n");
}

export function formatInputForConsole(
  date: Date,
  data: prompt.Properties
): string {
  const colorizeMark = (value: string) => {
    switch (value) {
      case "1":
        return chalk.rgb(150, 0, 0).bold(value);
      case "2":
        return chalk.rgb(255, 125, 125).bold(value);
      case "3":
        return chalk.rgb(255, 255, 75).bold(value);
      case "4":
        return chalk.rgb(125, 255, 125).bold(value);
      case "5":
        return chalk.rgb(0, 180, 0).bold(value);
    }
    return chalk.white(value);
  };

  return `${chalk.blue(formatDate(date))}\n`.concat(
    columnify(
      Object.keys(data).map((k) => ({
        project: chalk.yellow(k),
        mark: colorizeMark(data[k] as string),
      }))
    )
  );
}
