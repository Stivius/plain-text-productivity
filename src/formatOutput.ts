import chalk from 'chalk';
import columnify from 'columnify'

export function formatDataForFile(data) {
	return Object.keys(data).map((k) => {
		if (k !== 'Date') {
			return `${k}:${data[k]}`
		} else {
			return `${data[k]}`;
		}
	}).join('\n').concat('\n');
}

export function outputDataToConsole(data) {
	const colorizeMark = (value: string) => {
		switch (value) {
			case '1': return chalk.rgb(150, 0, 0).bold(value); 
			case '2': return chalk.rgb(255, 125, 125).bold(value); 
			case '3': return chalk.rgb(255, 255, 75).bold(value); 
			case '4': return chalk.rgb(125, 255, 125).bold(value); 
			case '5': return chalk.rgb(0, 180, 0,).bold(value); 
		}
		return chalk.white(value);
	};

	const DATE_KEY = 'Date';
	if (data.hasOwnProperty(DATE_KEY)) {
		console.log(`${chalk.blue(data[DATE_KEY])}`);
		delete data[DATE_KEY];
	}

	console.log(columnify(Object.keys(data).map((k) => ({
		project: chalk.yellow(k),
		mark: colorizeMark(data[k] as string),
	}))));
}

