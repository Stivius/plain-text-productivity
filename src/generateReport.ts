import chain, {filter, round, sumBy} from 'lodash'
import {FileData, ProductivityReport, ProductivityReportOptions} from './interfaces'

export function generateReport(data: FileData, options: ProductivityReportOptions): ProductivityReport {
    const groupedData = chain(data.records)
    .filter((r) => r.day >= options.from && r.day <= options.to)
    .map((r) => r.projects)
    .flatMap()
    .filter((p) => data.metadata.activeProjects.indexOf(p.name) !== -1)
    .groupBy((assesed) => assesed.name);
    const averageMarkByProject = groupedData.map((group, groupName) => {
        const groupSum = sumBy(group, (assesed) => assesed.mark);
        const filteredGroup = filter(group, (assesed) => assesed.mark !== undefined);
        return { name: groupName, mark: groupSum / filteredGroup.length };
    });
    const maxMarkFromAverages = averageMarkByProject.map(p => p.mark - 1).max();
    const items = averageMarkByProject.map((value) => {
        const mark = value.mark - 1;
        const MAX_MARK = 4;
        if (options.absolute) {
            return { name: value.name, productivity: round(mark / MAX_MARK, 4) };
        }
        return { name: value.name, productivity: round(mark / maxMarkFromAverages, 4) };
    }).value();
    return {
        items: items,
        options: options
    };
}
