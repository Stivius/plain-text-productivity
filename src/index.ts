import commandLineArgs from 'command-line-args';
import fs from 'fs';
import dateFormat from 'dateformat'
import {readRecordFromConsole} from './readRecordFromInput';
import chalk from 'chalk';
import {formatDataForFile, outputDataToConsole} from './formatOutput';
import {EstimatedProject, FileData, FILE_NAME, ProjectRecord} from './interfaces';
import {readRangeChoiceFromConsole} from './readRangeChoiceFromInput';
import {getRangeByType} from './getRangeByType';
import {FileParser} from './fileParser';
import {generateReport} from './generateReport';
import {outputReport} from './outputReport';

const commandsDefinitions = [
    { name: 'command', defaultOption: true },
]

const commands = commandLineArgs(commandsDefinitions, { stopAtFirstUnknown: true });

async function main() {
    try {
        const parser = new FileParser(FILE_NAME);
        const data: FileData = parser.parse();

        switch (commands.command) {
            // TODO: refactor output to console/file
            case 'add': {
                const enteredData = await readRecordFromConsole(data.metadata.activeProjects);
                const date = dateFormat(new Date(), "yyyy-mm-dd");
                console.log(chalk.blue(date));
                fs.appendFileSync(FILE_NAME, date.concat('\n', formatDataForFile(enteredData), '\n'));
                outputDataToConsole(enteredData);
                break;
            }

            case 'addp': {
                const enteredData = await readRecordFromConsole(data.metadata.activeProjects, true);
                fs.appendFileSync(FILE_NAME, formatDataForFile(enteredData).concat('\n'));
                outputDataToConsole(enteredData);
                break;
            }

            case 'lsp':
                const projects = data.records.map((r: ProjectRecord) => r.projects.map((p: EstimatedProject) => p.name));
            const uniqueProjects = Array.from(new Set(projects.flat()));
            const achivedProjects = uniqueProjects.filter((p) => !data.metadata.activeProjects.includes(p));
            console.log('All projects:');
            console.log(chalk.yellow(data.metadata.activeProjects.join('\n')));
            console.log(chalk.gray(achivedProjects.join('\n')));
            break;

            case 'lspa':
                console.log('Active projects:');
            console.log(chalk.yellow(data.metadata.activeProjects.join('\n')));
            break;

            case 'report': {
                const optionDefinitions = [
                    { name: 'absolute', type: Boolean, defaultValue: false },
                    { name: 'range-type', type: Number },
                    { name: 'from-date', type: String },
                    { name: 'to-date', type: String },
                ]
                const options = commandLineArgs(optionDefinitions, { argv: commands._unknown || [] });
                const getRange = (options: commandLineArgs.CommandLineOptions) => {
                    const rangeType = options['range-type'];
                    if (rangeType === undefined) {
                        return readRangeChoiceFromConsole();
                    } else {
                        // TODO: can be entered as string and converted to number
                        if (rangeType === 0) {
                            const from = new Date(options['from-date']);
                            const to = new Date(options['to-date']);
                            if (!from || !to) {
                                throw new Error('Date range is not specified or invalid');
                            }
                            return { from, to, choice: 0 };
                        }
                        return getRangeByType(rangeType);
                    }
                };

                const { from, to, choice } = await getRange(options);
                const productivityByProject = generateReport(data.records, from, to, options.absolute);
                outputReport(productivityByProject, choice, from, to, options.absolute);
                break;
            }
            default:
                console.error('Unknown command');
            break;
        }
    } catch (error) {
        console.error(error);
    }
}

main().catch((err) => console.log(err, 'Error while running'))
