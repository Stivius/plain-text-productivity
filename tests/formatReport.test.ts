import { formatReport } from "../src/formatReport";
import {
  ProductivityReportItem,
  ProductivityReportOptions,
} from "../src/interfaces";

function reportOptions(options: {
  absolute?: boolean;
  choice?: number;
}): ProductivityReportOptions {
  return {
    from: new Date("2022-07-01"),
    to: new Date("2022-07-04"),
    absolute: options.absolute === undefined ? true : options.absolute,
    choice: options.choice || 0,
  };
}

const REPORT_ITEMS: ProductivityReportItem[] = [
  { name: "zero", productivity: 0 },
  { name: "first", productivity: 0.1 },
  { name: "second", productivity: 0.2 },
  { name: "third", productivity: 0.3 },
  { name: "fourth", productivity: 0.4 },
  { name: "fifth", productivity: 0.5 },
  { name: "sixth", productivity: 0.6 },
  { name: "seventh", productivity: 0.7 },
  { name: "eight", productivity: 0.8 },
  { name: "ninth", productivity: 0.9 },
  { name: "tenth", productivity: 1 },
];

describe("formatReport", () => {
  test.each([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10])(
    "report with all choices",
    (choice: number) => {
      expect(
        formatReport({
          items: REPORT_ITEMS,
          options: reportOptions({ choice: choice }),
        })
      ).toMatchSnapshot();
    }
  );

  test.each([false, true])("report relativeness", (absolute: boolean) => {
    expect(
      formatReport({
        items: REPORT_ITEMS,
        options: reportOptions({ absolute: absolute }),
      })
    ).toMatchSnapshot();
  });
});
