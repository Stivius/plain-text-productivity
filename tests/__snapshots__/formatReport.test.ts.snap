// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatReport report relativeness 1`] = `
"Custom Relative Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report relativeness 2`] = `
"Custom Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 1`] = `
"Custom Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 2`] = `
"Daily Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 3`] = `
"Weekly Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 4`] = `
"Monthly Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 5`] = `
"Yearly Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 6`] = `
"7-days Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 7`] = `
"14-days Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 8`] = `
"30-days Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 9`] = `
"90-days Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 10`] = `
"180-days Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;

exports[`formatReport report with all choices 11`] = `
"365-days Absolute Report for range [34m2022-07-01[39m - [34m2022-07-04[39m
NAME    PRODUCTIVITY   
[33mzero[39m    [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  0.00%[22m[39m  
[33mfirst[39m   [38;2;150;0;0m[1mâ– â– â– â– â– â– â–  10.00%[22m[39m 
[33msecond[39m  [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  20.00%[22m[39m 
[33mthird[39m   [38;2;255;125;125m[1mâ– â– â– â– â– â– â–  30.00%[22m[39m 
[33mfourth[39m  [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  40.00%[22m[39m 
[33mfifth[39m   [38;2;255;255;75m[1mâ– â– â– â– â– â– â–  50.00%[22m[39m 
[33msixth[39m   [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  60.00%[22m[39m 
[33mseventh[39m [38;2;125;255;125m[1mâ– â– â– â– â– â– â–  70.00%[22m[39m 
[33meight[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  80.00%[22m[39m 
[33mninth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  90.00%[22m[39m 
[33mtenth[39m   [38;2;0;180;0m[1mâ– â– â– â– â– â– â–  100.00%[22m[39m"
`;
